lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2135,26 +2135,57 @@ namespace Js
     {
         charcount_t count = pThis->GetLength();
 
-        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
-        const wchar_t *inStr = pThis->GetString();
-        wchar_t *outStr = builder.DangerousGetWritableBuffer();
+        const wchar_t* inStr = pThis->GetString();
+        const wchar_t* i = inStr; 
+        const wchar_t* iLim = i + count;
+        
+        // first, try to see if the string needs recasing at all
+        while(i < iLim)
+        {
+            
+            // copy the current char into a temporary variable, then case it
+            wchar_t inChar = *i;
+            wchar_t outChar = *i;
+            CharUpperBuffW(&outChar, 1);
+            
+            // if the output is different from the input, the string needs recasing from here
+            if(outChar != inChar) { break; }
+            
+            // otherwise, we can keep going on
+            i++;
+            
+        }
+        
+        // if we reached the end of the string and didn't need recasing, we can return immediately
+        if(i == iLim)
+        { 
+            return pThis;
+        }
+        
+        // if not, we have to copy the first part of the string, then case the rest
+        charcount_t countToSkip = i - inStr;
+        charcount_t countToConvert = count - countToSkip; 
 
+        // start by copying the whole string in a new buffer
+        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
+        wchar_t* outStr = builder.DangerousGetWritableBuffer();
         wchar_t* outStrLim = outStr + count;
-        wchar_t *o = outStr;
+        wchar_t* o = outStr;
 
         while (o < outStrLim)
         {
             *o++ = *inStr++;
         }
-
+        
+        // then recase it
         if(toCase == ToUpper)
         {
 #if DBG
             DWORD converted =
 #endif
-                CharUpperBuffW(outStr, count);
+                CharUpperBuffW(outStr + countToSkip, countToConvert);
```

Comment:
@dilijev , how do you think, it is correct to use WinAPI CharUpperBuff for string case conversion?
It is a common approach that length of uppercase string will be equal to length of lowercase string, but it is wrong.
For example http://stackoverflow.com/a/20301720